### ISSUE TITLE
HA for RouterOS 7.x.x

### ISSUE THREAD

**Issue by AndreasMagdeburg** [svlsResearch/ha-mikrotik#23]
Has anyone gotten the HA scripts to work on the new RouterOS software version 7.1.2? 
In my tests in the GNS3 Testlab, the scripts ran without errors on the first router, but they didn't run on the second router. 
A $HASyncStandby does not work.

**Comment by nathanfaber** [comment]
I have not attempted it yet. They still haven't fixed a bug in 7 that I want (OpenVPN UDP) so I haven't pursued it. I'm not sure how much work it would take to get ha-mikrotik to work under 7. I was somewhat hoping they would introduce their own so we can put this project out of business.

**Comment by AndreasMagdeburg** [comment]
I hope you will try to get ha-mikrotik working (on 7.1.2 and newer) very soon, because HA is a very important thing.
Great work!

**Comment by AndreasMagdeburg** [comment]
You are my last resort. 
Unfortunately, both of the newly purchased routers CCR2004-1G-12S+2XS already have the v7.1.1 as factory installed version. 
I can't go back to 6.49.3 on the routers and I absolutely need the HA.

**Comment by nathanfaber** [comment]
> You are my last resort. Unfortunately, both of the newly purchased routers CCR2004-1G-12S+2XS already have the v7.1.1 as factory installed version. I can't go back to 6.49.3 on the routers and I absolutely need the HA.

Doesn't the CCR2004 support 6.x? Why can't you go back? I'm sorry, I don't have the bandwidth nor the desire to investigate v7 at this point in time.

**Comment by AndreasMagdeburg** [comment]
> Doesn't the CCR2004 support 6.x? Why can't you go back? I'm sorry, I don't have the bandwidth nor the desire to investigate v7 at this point in time.

I can't go back to 6.x because the factory installed version is 7.1.1.
You can't downgrade below factory installed version.

**Comment by nathanfaber** [comment]
> I can't go back to 6.x because the factory installed version is 7.1.1.
> You can't downgrade below factory installed version.

I see. Another case of Mikrotik doing what Mikrotik does and frustrating users. I'm sorry you have run into this, for now I can't look into v7 but maybe in the future. Perhaps send Mikrotik an email and let them know you would like HA/config sync built in. They take forever to make changes/fixes but maybe one day I can just kill this project.

**Comment by AndreasMagdeburg** [comment]
Yes, it is fustrating, that there is no way to downgrade below factory installed version.
I had a long and difficult conversation with the mikrotik support.
HA is important. It's a shame that such a feature doesn't officially exist yet.

Aber der Hersteller ist leider derjenige, welcher über die Implementation entscheidet.  :-(

**Comment by agungsaptohermawan** [comment]
> > I can't go back to 6.x because the factory installed version is 7.1.1.
> > You can't downgrade below factory installed version.
> 
> I see. Another case of Mikrotik doing what Mikrotik does and frustrating users. I'm sorry you have run into this, for now I can't look into v7 but maybe in the future. Perhaps send Mikrotik an email and let them know you would like HA/config sync built in. They take forever to make changes/fixes but maybe one day I can just kill this project.

are you providing managed service for HA on ROS v7? our company looking for it

**Comment by nathanfaber** [comment]
> > > I can't go back to 6.x because the factory installed version is 7.1.1.
> > > You can't downgrade below factory installed version.
> > 
> > 
> > I see. Another case of Mikrotik doing what Mikrotik does and frustrating users. I'm sorry you have run into this, for now I can't look into v7 but maybe in the future. Perhaps send Mikrotik an email and let them know you would like HA/config sync built in. They take forever to make changes/fixes but maybe one day I can just kill this project.
> 
> are you providing managed service for HA on ROS v7? our company looking for it

Unfortunately not. I am not currently supporting HA for v7.

**Comment by agungsaptohermawan** [comment]
> > > I see. Another case of Mikrotik doing what Mikrotik does and frustrating users. I'm sorry you have run into this, for now I can't look into v7 but maybe in the future. Perhaps send Mikrotik an email and let them know you would like HA/config sync built in. They take forever to make changes/fixes but maybe one day I can just kill this project.
> > 
> > 
> > are you providing managed service for HA on ROS v7? our company looking for it
> 
> Unfortunately not. I am not currently supporting HA for v7.

Well, big and many thanks for your script and support. its work perfectly on our production env

**Comment by botts99** [comment]
Hello, I was able to sucessfully get this to work on a pair of hexs for testing using 7.1.2.  The main issue I seen was when using switchrole it would not fully switch as it was never able to ping the other standby unit.

I looked at some older firmware the main issue was when using the ping command even with a count of 1 it would return the status line.  In 7.1.2 it would not return the line until I increased the count from 1 to 4 in the ping command.  Once I fixed that the devices would rotate when powered cylced correctly.

As always test this yourself before production as I did not fully test every featue.



sent=4 received=4 packet-loss=0% min-rtt=632us avg-rtt=652us max-rtt=701us

add name=ha_switchrole_new owner=admin policy=ftp,reboot,read,write,policy,test,password,sniff,sensitive source=":global isMaster\
	\n:global haAddressOther\
	\n:global haInterface\
	\n:global haInterfaceLogical\
	\n:global haPreferMac\
	\n\
	\n:local haPingInterface\
	\n:if ([:typeof \$haInterfaceLogical] = \"nothing\") do={\
	\n   :set haPingInterface \"\$haInterface\"\
	\n} else={\
	\n   :set haPingInterface \"\$haInterfaceLogical\"\
	\n}\
	\n\
	\n:put \"Using ping interface \$haPingInterface\"\
	\n\
	\n:if (([:typeof \$haPreferMac] != \"nothing\") && (\$haPreferMac != \"\")) do={\
	\n   :put \"You are using haPreferMac (\$haPreferMac) - switch role does not make sense.\"\
	\n   return 0\
	\n}\
	\n\
	\n:if (\$isMaster) do={\
	\n   :put \"I am master - switching role\"\
	\n   /system script run [find name=\"ha_pushbackup\"]\
	\n   :delay 5\
	\n   :local haWaitCount 0\
	\n   while ([/ping \$haAddressOther count=4 interface=\$haPingInterface ttl=1]  = 0) do={\
	\n      :set haWaitCount (\$haWaitCount+1)\
	\n      :put \"Still waiting for standby \$haWaitCount...\"\
	\n      :delay 1\
	\n   }\
	\n   :put \"Standby available \$haWaitCount...delaying 5s\"\
	\n   /delay 5\
	\n   :if (\$isMaster && [/ping \$haAddressOther count=4 interface=\$haPingInterface ttl=1]  >= 1) do={\
	\n      :put \"REBOOTING MYSELF\"\
	\n      :execute \"/system reboot\"\
	\n   } else={\
	\n      :put \"NOT REBOOTING MYSELF! SLAVE IS NOT UP OR I AM NOT MASTER!\"\
	\n   }\
	\n} else={\
	\n   :put \"I am NOT master - nothing to do\"\
	\n}\
`

Another issue I found was when trying to setup the standy.   The file would get removed when you reboot and erase the config.

I didnt run these 2 commands 

**/delay 2
/file print file=HA_bootstrap.rsc**

Ran this and verified the file was correct copied to the flash and a reset system config from winbox and chose the rsc file from the flash. Once it booted I removed that file from flash and was able to sync the standy.

**/file set [find name=HA_bootstrap.rsc] contents=":local haBootstrapOK false; :while (!\$haBootstrapOK) do={:do { /ip address add address=\"169.254.23.2/24\" interface=ether3; /user add name=ha group=full password=\"Mypassword!\"; :set haBootstrapOK true;} on-error={/log warning \"ha_startup: 0.0 B bootstrap failed...waiting\"; :delay 5};}"**


#Maybe try: /tool mac-telnet xx:xx:xx:xx:xx:xx
###PASTE THIS ON THE OTHER DEVICE - YOUR CONFIG WILL BE RESET AND LOST!!!###
:global mac [[/interface ethernet get [find default-name="ether3"] orig-mac-address]]
:if ($mac != "xx:xx:xx:xx:xx:xx" and $mac != "xx:xx:xx:xx:xx:xx") do={
   :error "Interface ether3 MAC $mac does not match (A=xx:xx:xx:xx:xx:xx or B=xx:xx:xx:xx:xx:xx) - please check config\r\nUse orig-mac address!"
}
:if ([:len [/file find name=HA_backup_beforeHA.backup]] = 0) do={
   /system backup save name=HA_backup_beforeHA dont-encrypt=yes
   /export file=HA_backup_beforeHA.rsc
}
/export file=HA_bootstrap.rsc
/delay 2
/file print file=HA_bootstrap.rsc
/file set [find name=HA_bootstrap.rsc] contents=":local haBootstrapOK false; :while (!\$haBootstrapOK) do={:do { /ip address add address=\"169.254.23.2/24\" interface=ether3; /user add name=ha group=full password=\"Mypassword!\"; :set haBootstrapOK true;} on-error={/log warning \"ha_startup: 0.0 B bootstrap failed...waiting\"; :delay 5};}"
/system reset-configuration no-defaults=yes keep-users=no skip-backup=yes run-after-reset=HA_bootstrap.rsc
###END OF PASTE FOR OTHER DEVICE###
###


Anyways hope this helps some people if the need it.

**Comment by timeport0** [comment]
I just tried this on a set of CRS504 with 7.10rc5 and everything so far is going almost seamlessly with zero modifications to the scripts

Had a very small issue where the port on the active was in a bridge, so it didn't get added to the new bridge created by the script. Once I fixed that everything synced up perfectly. Haven't fully tested failover, but config changes are indeed syncing.

Will be doing additional testing and will report back

**Comment by rdy28** [comment]
Hi is anyone tested the HA in the latest version now in 2024?

**Comment by nathanfaber** [comment]
> Hi is anyone tested the HA in the latest version now in 2024?

Yes, I have been running and testing v7 using this branch: https://github.com/svlsResearch/ha-mikrotik/tree/v7-test

**Comment by rdy28** [comment]
> > Hi is anyone tested the HA in the latest version now in 2024?
> 
> Yes, I have been running and testing v7 using this branch: https://github.com/svlsResearch/ha-mikrotik/tree/v7-test

Thank you so much!

**Comment by devopslogical** [comment]
v7-test branch works fine up until RouterOS version 7.12.2

If you upgrade to 7.13 or higher, you get some weird errors when scripts are saving files. My guess is from RouterOS changes regarding filesystem.

Can someone take a look at it?

**Comment by dennes544** [comment]
@devopslogical Running on 7.14.3 for over a 6 months in prod, everything seems fine

**Comment by MetUys** [comment]
I'm also running it on v7 (v7.15.2) also over 6months in production, all good so far.

**Comment by devopslogical** [comment]
@dennes544 @MetUys Thank you for the response.
I think that issues that I encountered with 7.13 were resolved in later versions 7.13.1 ..  7.13.5 since it had lots if hotfixes.

Managed to upgrade to latest stable 7.18.2 and it is looking good.

**Comment by DeZ641** [comment]
Hello, I have the following situation: 
there is a task in the scheduler called ha_checkchanges that runs every 10 minutes, and during each run, files are copied to the Standby device. The field system/script/environment/haCheckStandbyVer is always empty. Is this expected behavior if no changes have been made? 

How can I ensure that some values appear in the haCheckStandbyVer field? Thank you.

**Comment by MetUys** [comment]
**Update:** I have done a clean setup with two CCR2116-12G-4S+ (arm64) netinstalled to v7.20.8 (latest long term version at the time of writing).

I had some teething issues, which was related to the console window size that messed with the $HAInstall output that then resulted in a malformed HA_bootstrap.rsc script file. (This is not a HA-Mikrotik issue but a user side consideration). After figuring that out and adjusting, everything went smoothly.

Without too much config deployed I have tested the syncing and roleswitching (graceful + "unplanned" events), all went well.

I will now load up more configurations onto the pair and report back.

**Comment by MetUys** [comment]
I have been running the HA on RouterOS v7.20.8 since April with no issues.
With the vulnerability drama unfolding, I just wanted to check if anyone has done the upgrade to RouterOS v7.23.5 on an HA pair of Mikrotiks? (or at least any noted issues for things newer than v7.20.8 to present).
